README.md
# ***Fusion360  Two-turn Möbius strip***
This is an add-in for the Autodesk software "Fusion360".

Create a two-turn Möbius strip.

![Alt text](./resources/result.png)

***
## Installation :
To install, download/extract the zip file and register the "Two-turn_Mobius_strip" folder as an add-in.

To register an add-in,Please follow the instructions [here](https://knowledge.autodesk.com/support/fusion-360/troubleshooting/caas/sfdcarticles/sfdcarticles/How-to-install-an-ADD-IN-and-Script-in-Fusion-360.html).

## Usage :
+ Launch the add-in. The command will be added within the "Create" tab of "Solids".

    ![Alt text](./resources/tool_panel_eng.png)
+ Execute the command.

+ A dialog box will appear. Set each item.

    ![Alt text](./resources/dialog_eng.png)
    + Center Point : The center point of the Möbius strip to be created. Specify the sketch point.
    + Twist Count : Specify the number of times you want to twist the strip (one time is 180°) between two revolutions.
        + Due to the nature of the Möbius strip, only odd numbers are available.
        + The limit is in the range of 1~29 times.
    + Hole Count : 2Specify the number of holes to avoid interference when rotating.
        + For the convenience of creating an even arrangement of holes, only odd numbers are available.
        + It is limited to 13~49 pieces. 
        + Depending on the number of holes, the overall size will be changed.
    + Thickness : Enter the thickness dimensions of the Möbius strip.
        + Excessive thickness may result in an error.
    + Information : If the set value is not correct, the reason for the cause is displayed.

+ By pressing the OK button, a two-turn Möbius strip will be created in the new component.

## Constraints：
- Due to the complexity of the shape, it will take some time to create.
- The preview is shown without the hole in order to show it in a short time.

## Issue：
- without

## Action :
Confirmed in the following environment.
 + Fusion360 Ver2.0.9719
 + Windows10 64bit Pro , Home

## License :
MIT

## Acknowledgments :
+ Thank you everyone in the [Japanese forum.](https://forums.autodesk.com/t5/fusion-360-ri-ben-yu/bd-p/707)

***
***
以下、日本語です。

これは、オートデスクソフトウェア「Fusion360」のアドインです。

2回転のメビウスの帯を作成します。

![Alt text](./resources/result.png)

***
## インストール :
インストールする際は、zipファイルをダウンロード/展開後 "Two-turn_Mobius_strip" フォルダをアドインとして登録してください。

アドインの登録は、[こちら](https://knowledge.autodesk.com/ja/support/fusion-360/troubleshooting/caas/sfdcarticles/sfdcarticles/JPN/How-to-install-an-ADD-IN-and-Script-in-Fusion-360.html)の手順に従ってください。

## 使用法 :
+ アドインを起動。"ソリッド" の "作成" タブ内にコマンドが追加されます。

    ![Alt text](./resources/tool_panel_jpn.png)
+ コマンドを実行。

+ ダイアログが表示されます。各項目を設定して下さい。

    ![Alt text](./resources/dialog_jpn.png)
    + 中心点 : 作成するメビウスの帯の中心点。スケッチ点を指定してください。
    + ねじり回数 : 2回転する間に帯をねじる回数(1回は180°)を指定してください。
        + メビウスの帯の性質上、奇数のみです。
        + 1~29回の範囲で制限しています。
    + 穴の数 : 2回転する際、干渉を回避するための穴の数を指定してください。
        + 均等な穴の配置を作成する都合上、奇数のみです。
        + 13~49個の範囲で制限しています。 
        + 穴の数により、全体の大きさが変更されます。
    + 厚さ : メビウスの帯の厚さ寸法を入力してください。
        + 過剰な厚みにするとエラーとなる可能性が有ります。
    + 情報 : 設定値が正しくない場合に、原因となる理由を表示します。


+ OKボタンを押すことで、新たなコンポーネントに2回転メビウスの帯が作成されます。

## 制約 ：
- 形状が複雑な為、作成時に程々の時間がかかります。
- プレビューは短時間で表示させるために、穴の開いていない状態を表示させています。

## 問題 ：
- 無し

## アクション :
以下の環境で確認しています。
 + Fusion360 Ver2.0.9719
 + Windows10 64bit Pro , Home

## ライセンス :
MIT

## 謝辞 :
+ [日本語フォーラム](https://forums.autodesk.com/t5/fusion-360-ri-ben-yu/bd-p/707)の皆さん、ありがとう。
+ 上記の英語で不適切な部分をして頂けると助かります。

CHANGELOG.md
# Change of Fusion360  Two-turn Möbius strip

## 0.0.2
+ fix bug at startup.

## 0.0.1
+ release

Two-turn Mobius strip/Two-turn Mobius strip.py
#Author-kantoku
#Description-Two-turn Mobius strip
#Fusion360API Python

import adsk.core
import traceback

try:
    from . import config
    from .apper import apper

    from .commands.MobiusStripCore import MobiusStripCore
    from .commands.ktkLanguageMessage import LangMsg

    msgDict = {
        'Two-turn Mobius strip' : '2回転メビウスの帯',
        'Create a Two-turn Mobius strip' : '2回転メビウスの帯を作成'
    }
    lm = LangMsg(msgDict, adsk.core.UserLanguages.JapaneseLanguage)

    # Create our addin definition object
    my_addin = apper.FusionApp(config.app_name, config.company_name, False)
    my_addin.root_path = config.app_path

    my_addin.add_command(
        lm.sLng('Two-turn Mobius strip'),
        MobiusStripCore,
        {
            'cmd_description': lm.sLng('Create a Two-turn Mobius strip'),
            'cmd_id': 'two-turn_Mobius_strip',
            'workspace': 'FusionSolidEnvironment',
            'toolbar_panel_id': 'SolidCreatePanel',
            'cmd_resources': 'TwoTurnMobiusStrip',
            'command_visible': True,
            'command_promoted': False,
            'create_feature': False,
        }
    )

except:
    app = adsk.core.Application.get()
    ui = app.userInterface
    if ui:
        ui.messageBox('Initialization: {}'.format(traceback.format_exc()))


def run(context):
    my_addin.run_app()


def stop(context):
    my_addin.stop_app()


Two-turn Möbius strip/Two-turn Möbius strip.py
#Author-kantoku
#Description-Two-turn Möbius strip
#Fusion360API Python

import adsk.core
import traceback

try:
    from . import config
    from .apper import apper

    from .commands.MöbiusStripCore import MöbiusStripCore
    from .commands.ktkLanguageMessage import LangMsg

    msgDict = {
        'Two-turn Möbius strip' : '2回転メビウスの帯',
        'Create a Two-turn Möbius strip' : '2回転メビウスの帯を作成'
    }
    lm = LangMsg(msgDict, adsk.core.UserLanguages.JapaneseLanguage)

    # Create our addin definition object
    my_addin = apper.FusionApp(config.app_name, config.company_name, False)
    my_addin.root_path = config.app_path

    my_addin.add_command(
        lm.sLng('Two-turn Möbius strip'),
        MöbiusStripCore,
        {
            'cmd_description': lm.sLng('Create a Two-turn Möbius strip'),
            'cmd_id': 'two-turn_Möbius_strip',
            'workspace': 'FusionSolidEnvironment',
            'toolbar_panel_id': 'SolidCreatePanel',
            'cmd_resources': 'TwoTurnMöbiusStrip',
            'command_visible': True,
            'command_promoted': False,
            'create_feature': False,
        }
    )

except:
    app = adsk.core.Application.get()
    ui = app.userInterface
    if ui:
        ui.messageBox('Initialization: {}'.format(traceback.format_exc()))


def run(context):
    my_addin.run_app()


def stop(context):
    my_addin.stop_app()


Two-turn Mobius strip/commands/MobiusStripCore.py
#Author-kantoku
#Description-Two-turn Mobius strip
#Fusion360API Python

import adsk.core
import adsk.fusion

from ..apper import apper
from .. import config
from .ktkCmdInputHelper import SelectionCommandInputHelper, IntegerSpinnerCommandInputHelper
from .ktkCmdInputHelper import TextBoxCommandInputHelper, ValueCommandInputHelper
from .MobiusStripFactry import MobiusStripFactry
from .ktkLanguageMessage import LangMsg

msgDict = {
    'Center Point' : '中心点',
    'Select Sketch Point' : 'スケッチ点を選択',
    'Twist Count' : 'ねじる回数',
    'Hole Count' : '穴の数',
    'Thickness' : '厚さ',
    'Information' : '情報',
    'is out of scope!' : 'が範囲外です!',
    'is an odd number only!' : 'は奇数のみです!',
    'should be a positive value!' : '0以上にして下さい!'
}

lm = LangMsg(msgDict, adsk.core.UserLanguages.JapaneseLanguage)

_point = SelectionCommandInputHelper(
    'pointIpt',
    lm.sLng('Center Point'),
    lm.sLng('Select Sketch Point'),
    ['SketchPoints'])

_twist = IntegerSpinnerCommandInputHelper(
    'twistIpt',
    lm.sLng('Twist Count'),
    1,
    29,
    2,
    1)

_hole = IntegerSpinnerCommandInputHelper(
    'holeIpt',
    lm.sLng('Hole Count'),
    13,
    49,
    2,
    13)

_thick = ValueCommandInputHelper(
    'thickIpt',
    lm.sLng('Thickness'))

_info = TextBoxCommandInputHelper(
    'infoTxt',
    lm.sLng('Information'),
    '',
    10,
    True)

class MobiusStripCore(apper.Fusion360CommandBase):

    def on_preview(self, command, inputs, args, input_values):
        global _point, _twist, _hole, _thick

        checkLst = [
            _twist.isRange(),
            _twist.isOdd(),
            _hole.isRange(),
            _hole.isOdd()
        ]

        if not all(checkLst):
            return

        MobiusStripFactry.previewMobiusStrip(
            _point.obj.selection(0).entity,
            _twist.obj.value,
            _hole.obj.value
        )

    def on_destroy(self, command, inputs, reason, input_values):
        pass

    def on_input_changed(self, command, inputs, changed_input, input_values):
        global _point, _twist, _hole, _thick, _info

        msg = []

        if not _twist.isRange():
            msg.append(f'{_twist.name} ' + lm.sLng('is out of scope!') + 
                f'({_twist.obj.minimumValue} - {_twist.obj.maximumValue})')

        if not _twist.isOdd():
            msg.append(f'{_twist.name} ' + lm.sLng('is an odd number only!'))


        if not _hole.isRange():
            msg.append(f'{_hole.name} ' + lm.sLng('is out of scope!') + 
                f'({_hole.obj.minimumValue} - {_hole.obj.maximumValue})')

        if not _hole.isOdd():
            msg.append(f'{_hole.name} ' + lm.sLng('is an odd number only!'))


        if not (_thick.obj.value > 0):
            msg.append(f'{_thick.name} ' + lm.sLng('should be a positive value!'))


        if len(msg) > 0:
            _info.obj.text = '\n'.join(msg)
        else:
            _info.obj.text = ''


        command.doExecutePreview()

    def on_execute(self, command, inputs, args, input_values):
        global _point, _twist, _hole, _thick
        thick = _thick.obj.value * 0.5
        MobiusStripFactry.createMobiusStrip(
            _point.obj.selection(0).entity,
            _twist.obj.value,
            _hole.obj.value,
            thick
        )

        _twist.initialValue =  _twist.obj.value
        _hole.initialValue = _hole.obj.value
        _thick.initialValue = adsk.core.ValueInput.createByString(_thick.obj.expression)

    def on_create(self, command, inputs):
        global _point, _twist, _hole, _thick, _info
        _point.register(inputs)
        _twist.register(inputs)
        _hole.register(inputs)
        _thick.register(inputs)
        _info.register(inputs)

    # expansions
    def on_validate(self, command, inputs, args, input_values):
        global _point, _twist, _hole, _thick, _info

        checkLst = [
            _twist.isOdd(),
            _hole.isOdd(),
            _thick.obj.value > 0
        ]

        if all(checkLst):
            args.areInputsValid = True
        else:
            args.areInputsValid = False

    # expansion
    def on_preselect(self, command, inputs, args, pre_input, input_values):
        pass

Two-turn Mobius strip/commands/MobiusStripFactry.py
#Author-kantoku
#Description-Two-turn Mobius strip
#Fusion360API Python

import adsk.core
import adsk.fusion
import traceback
import math
from typing import List

_debug = False

def run(context):
    ui = None
    try:
        app = adsk.core.Application.get()
        ui = app.userInterface

        MobiusStripFactry.unitTest()

    except:
        if ui:
            ui.messageBox('Failed:\n{}'.format(traceback.format_exc()))

class MobiusStripFactry:

    @staticmethod
    def createMobiusStrip(
        center :adsk.fusion.SketchPoint,
        twistCount :int,
        holeCount :int,
        thick :float
        ) -> adsk.fusion.BRepBody:

        # numerical calculation
        rotations = [0, 240, 480]

        # base point
        basePnt = adsk.core.Point3D.create(holeCount * 0.25, 0, 0)

        # sweep
        sweeps = [MobiusStripFactry._initSweep(basePnt, twistCount, ang) for ang in rotations]

        if _debug:
            trims = sweeps
        else:
            # sphere
            spheres = [MobiusStripFactry._initSphere(basePnt, ang, holeCount) for ang in rotations]

            # trim
            trims = [MobiusStripFactry._initTrim(sw,sps) for sw, sps in zip(sweeps, spheres)]

        # create BaseFeature
        baseBodies = MobiusStripFactry._createBaseFeature(trims)

        # get Matrix
        sktMat :adsk.core.Matrix3D = MobiusStripFactry._getSketchPointMatrix(center)
        mat :adsk.core.Matrix3D = sktMat.copy()

        # create occurrence
        app :adsk.core.Application = adsk.core.Application.get()
        des :adsk.fusion.Design = app.activeProduct
        root :adsk.fusion.Component = des.rootComponent
        occ :adsk.fusion.Occurrence = root.occurrences.addNewComponent(
            adsk.core.Matrix3D.create())
        occ.component.name = 'Two-turn Mobius strip'

        # thickenBody
        MobiusStripFactry._createThickenBody(
            baseBodies,
            adsk.core.ValueInput.createByReal(thick * 0.5),
            occ.component,
            mat)

        # body rename
        bodies :adsk.fusion.BRepBodies = occ.component.bRepBodies
        if bodies.count > 0:
            bodies[0].name = f'twist{twistCount}_hole{holeCount}'
            return bodies[0]
        else:
            return None

    @staticmethod
    def previewMobiusStrip(
        center :adsk.fusion.SketchPoint,
        twistCount :int,
        holeCount :int,
        ) -> List[adsk.fusion.BRepBody]:

        # numerical calculation
        rotations = [0, 240, 480]

        # base point
        basePnt = adsk.core.Point3D.create(holeCount * 0.25, 0, 0)

        # sweep
        sweeps = [MobiusStripFactry._initSweep(basePnt, twistCount, ang) for ang in rotations]

        # get mat
        sktMat :adsk.core.Matrix3D = MobiusStripFactry._getSketchPointMatrix(center)
        mat :adsk.core.Matrix3D = sktMat.copy()

        # create BaseFeature
        baseBodies = MobiusStripFactry._createBaseFeature(sweeps, mat)

        return baseBodies


    # sweep
    @staticmethod
    def _initSweep(
        basePnt :adsk.core.Point3D,
        twistCount :int,
        posAng :float
        ) -> adsk.fusion.BRepBody:

        app = adsk.core.Application.get()
        des :adsk.fusion.Design = app.activeProduct
        root :adsk.fusion.Component = des.rootComponent

        pnt3D = adsk.core.Point3D
        vec3D = adsk.core.Vector3D

        width = 4.0
        twistAllAng = 180 * twistCount
        twistStartAng = twistAllAng * (1/3) * (posAng / 240)
        twistAngle = twistAllAng * (1/3)

        dustBox = []

        mat :adsk.core.Matrix3D = adsk.core.Matrix3D.create()
        mat.setToRotation(
            math.radians(posAng),
            vec3D.create(0,0,1),
            pnt3D.create(0,0,0))

        # prof coordinate
        profMin :adsk.core.Point3D = pnt3D.create(
            basePnt.x + (-width * 0.5 * math.cos(math.radians(-twistStartAng))),
            0,
            -width * 0.5 * math.sin(math.radians(-twistStartAng)))
        profMin.transformBy(mat)

        profMax :adsk.core.Point3D = pnt3D.create(
            basePnt.x + (width * 0.5 * math.cos(math.radians(-twistStartAng))),
            0,
            width * 0.5 * math.sin(math.radians(-twistStartAng)))
        profMax.transformBy(mat)

        # sketch
        skt :adsk.fusion.Sketch = root.sketches.addWithoutEdges(
            root.xYConstructionPlane)
        dustBox.append(skt)
        skt.isLightBulbOn = False

        # prof
        lines :adsk.fusion.SketchLines = skt.sketchCurves.sketchLines
        line = lines.addByTwoPoints(profMin, profMax)

        prof :adsk.fusion.Profile = root.createOpenProfile(line)

        # path
        pathPnt :adsk.core.Point3D = basePnt.copy()
        pathPnt.transformBy(mat)
        arcs :adsk.fusion.sketchArcs = skt.sketchCurves.sketchArcs
        guide :adsk.fusion.SketchArc = arcs.addByCenterStartSweep(
            pnt3D.create(0,0,0),
            pathPnt,
            math.radians(240))

        NOCHAINED = adsk.fusion.ChainedCurveOptions.noChainedCurves
        path = adsk.fusion.Path.create(guide, NOCHAINED)

        # sweep
        swpFeats :adsk.fusion.SweepFeatures = root.features.sweepFeatures
        NEWBODY = adsk.fusion.FeatureOperations.NewBodyFeatureOperation
        valIpt = adsk.core.ValueInput

        twistVal = valIpt.createByReal(math.radians(twistAngle))

        swpIpt :adsk.fusion.SweepFeatureInput = swpFeats.createInput(prof, path, NEWBODY)
        swpIpt.twistAngle = twistVal
        swpIpt.isSolid = False

        swpFeat :adsk.fusion.SweepFeature = swpFeats.add(swpIpt)
        dustBox.append(swpFeat)

        tmpMgr = adsk.fusion.TemporaryBRepManager.get()
        clone :adsk.fusion.BRepBody = tmpMgr.copy(swpFeat.bodies[0])

        # delete skt,sweep
        [d.deleteMe() for d in dustBox[::-1]]

        return clone


    # sphere
    @staticmethod
    def _initSphere(
        basePnt :adsk.core.Point3D,
        posAng :float,
        holeCount :int,
        radius :float = 1.4
        ) -> List[adsk.fusion.BRepBody]:

        pnt3D = adsk.core.Point3D
        vec3D = adsk.core.Vector3D
        mat3D = adsk.core.Matrix3D

        unitAng = 720 / holeCount
        spCount = int(240 // unitAng) + 2
        vecZ = vec3D.create(0,0,1)
        pntZero = pnt3D.create(0,0,0)
        startAng = int(posAng / unitAng) * unitAng

        tmpMgr = adsk.fusion.TemporaryBRepManager.get()
        baseSp = tmpMgr.createSphere(basePnt, radius)
        spheres = []
        for idx in range(spCount):
            clone = tmpMgr.copy(baseSp)
            mat = mat3D.create()
            mat.setToRotation(math.radians(unitAng * idx + startAng), vecZ, pntZero)
            tmpMgr.transform(clone, mat)
            spheres.append(clone)

        return spheres


    # trim
    @staticmethod
    def _initTrim(
        sw :adsk.fusion.BRepBody,
        sps :List[adsk.fusion.BRepBody]
        ) -> adsk.fusion.BRepBody:

        tmpMgr = adsk.fusion.TemporaryBRepManager.get()

        clone = tmpMgr.copy(sw)
        boolDiff = adsk.fusion.BooleanTypes.DifferenceBooleanType
        for sp in sps:
            tool = tmpMgr.copy(sp.faces[0])
            tmpMgr.booleanOperation(clone,tool,boolDiff)

        return clone


    # insert Bodies
    @staticmethod
    def _createBaseFeature(
        surfs :List[adsk.fusion.BRepBody],
        mat :adsk.core.Matrix3D = None
        ) -> List[adsk.fusion.BRepBody]:

        app = adsk.core.Application.get()
        des :adsk.fusion.Design = app.activeProduct
        comp :adsk.fusion.Component = des.rootComponent



        matZero :adsk.core.Matrix3D = adsk.core.Matrix3D.create()
        if not mat:
            mat = matZero.copy()

        # transform
        if not matZero.isEqualTo(mat):
            tmpMgr = adsk.fusion.TemporaryBRepManager.get()
            [tmpMgr.transform(surf, mat) for surf in surfs]

        # BaseFeature
        baseFeats :adsk.fusion.BaseFeatures = comp.features.baseFeatures
        baseFeat = adsk.fusion.BaseFeature.cast(None)
        bodyLst = []
        try:
            baseFeat = baseFeats.add()
            baseFeat.startEdit()
            for surf in surfs:
                bodyLst.append(comp.bRepBodies.add(surf, baseFeat))
        except:
            pass
        finally:
            baseFeat.finishEdit()

        return bodyLst

    # get Matrix
    @staticmethod
    def _getSketchPointMatrix(
        sketchPnt :adsk.fusion.SketchPoint
        ) -> adsk.core.Matrix3D:

        try:
            # sketch point
            skt :adsk.fusion.Sketch = sketchPnt.parentSketch
            mat :adsk.core.Matrix3D = skt.transform
            _, xAxis, yAxis, zAxis = mat.getAsCoordinateSystem()
            mat.setWithCoordinateSystem(sketchPnt.worldGeometry, xAxis, yAxis, zAxis)

            if _debug :
                dumpMsg(f'skt_mat:{mat.asArray()}' )

            return mat
        except:
            # root
            return adsk.core.Matrix3D.create()

    # get Matrix -non use
    @staticmethod
    def _getOccMatrixFromSketchPoint(
        sketchPnt :adsk.fusion.SketchPoint
        ) -> adsk.core.Matrix3D:

        try:
            # occ
            skt :adsk.fusion.Sketch = sketchPnt.parentSketch
            occ :adsk.fusion.Occurrence = skt.assemblyContext
            des = adsk.fusion.Design.cast(occ.component.parentDesign)
            root = des.rootComponent

            mat = adsk.core.Matrix3D.create()
            occ_names = occ.fullPathName.split('+')
            if _debug:
                dumpMsg(f'occ_name:{occ.fullPathName}' )
            occs = [root.allOccurrences.itemByName(name) 
                        for name in occ_names]
            mat3ds = [occ.transform for occ in occs]
            mat3ds.reverse() #important!!
            for mat3d in mat3ds:
                mat.transformBy(mat3d)

            if _debug :
                dumpMsg(f'occ_mat:{mat.asArray()}' )

            return mat
        except:
            # root
            return adsk.core.Matrix3D.create()


    @staticmethod
    def _createThickenBody(
        bodyList :List[adsk.fusion.BRepBody],
        value :adsk.core.ValueInput,
        comp :adsk.fusion.Component,
        mat :adsk.core.Matrix3D = None
        ) -> adsk.fusion.BRepBody:

        # ObjectCollection
        objs = adsk.core.ObjectCollection.create()
        [objs.add(body.faces[0]) for body in bodyList]

        # thickenFeature
        thickFeats :adsk.fusion.ThickenFeatures  = comp.features.thickenFeatures
        NEWBODY = adsk.fusion.FeatureOperations.NewBodyFeatureOperation
        thickFeat = thickFeats.add(thickFeats.createInput(objs, value, True, NEWBODY))

        # clone
        tmpMgr = adsk.fusion.TemporaryBRepManager.get()
        clone :adsk.fusion.BRepBody = tmpMgr.copy(thickFeat.bodies[0])

        # get Matrix & transform
        matZero :adsk.core.Matrix3D = adsk.core.Matrix3D.create()
        if not mat:
            mat = matZero.copy()

        if not matZero.isEqualTo(mat):
            tmpMgr.transform(clone, mat)

        # BaseFeature
        baseFeats :adsk.fusion.BaseFeatures = comp.features.baseFeatures
        baseFeat = adsk.fusion.BaseFeature.cast(None)
        body = adsk.fusion.BRepBody.cast(None)

        try:
            baseFeat = baseFeats.add()
            baseFeat.startEdit()
            body = comp.bRepBodies.add(clone, baseFeat)
        except:
            pass
        finally:
            baseFeat.finishEdit()

        bodyList[0].baseFeature.deleteMe()
        thickFeat.deleteMe()

        return body


    @staticmethod
    def unitTest():
        import random
        ui = None

        try:
            app = adsk.core.Application.get()
            app.documents.add(adsk.core.DocumentTypes.FusionDesignDocumentType)
            ui = app.userInterface
            des :adsk.fusion.Design = app.activeProduct
            root :adsk.fusion.Component = des.rootComponent

            occ = root.occurrences.addNewComponent(adsk.core.Matrix3D.create())
            comps = [root, occ.component]

            pnt3D = adsk.core.Point3D

            for comp in comps:
                skt :adsk.fusion.Sketch = comp.sketches.add(comp.xYConstructionPlane)
                pnts :adsk.fusion.SketchPoints = skt.sketchPoints
                
                pntLst = []
                for _ in range(2):
                    pntLst.append(
                        pnts.add(
                            pnt3D.create(random.uniform(50, -50), random.uniform(50, -50), 0)))

                MobiusStripFactry.previewMobiusStrip(pntLst[0], 5, 23)
                app.activeViewport.refresh()
                adsk.doEvents()

                MobiusStripFactry.createMobiusStrip(pntLst[1], 1, 17, 0.05)
                app.activeViewport.refresh()
                adsk.doEvents()
        except:
            if ui:
                ui.messageBox('Failed:\n{}'.format(traceback.format_exc()))


def dumpMsg(msg :str):
    adsk.core.Application.get().userInterface.palettes.itemById('TextCommands').writeText(str(msg))

Two-turn Möbius strip/commands/MöbiusStripCore.py
#Author-kantoku
#Description-Two-turn Möbius strip
#Fusion360API Python

import adsk.core
import adsk.fusion

from ..apper import apper
from .. import config
from .ktkCmdInputHelper import SelectionCommandInputHelper, IntegerSpinnerCommandInputHelper
from .ktkCmdInputHelper import TextBoxCommandInputHelper, ValueCommandInputHelper
from .MöbiusStripFactry import MöbiusStripFactry
from .ktkLanguageMessage import LangMsg

msgDict = {
    'Center Point' : '中心点',
    'Select Sketch Point' : 'スケッチ点を選択',
    'Twist Count' : 'ねじる回数',
    'Hole Count' : '穴の数',
    'Thickness' : '厚さ',
    'Information' : '情報',
    'is out of scope!' : 'が範囲外です!',
    'is an odd number only!' : 'は奇数のみです!',
    'should be a positive value!' : '0以上にして下さい!'
}

lm = LangMsg(msgDict, adsk.core.UserLanguages.JapaneseLanguage)

_point = SelectionCommandInputHelper(
    'pointIpt',
    lm.sLng('Center Point'),
    lm.sLng('Select Sketch Point'),
    ['SketchPoints'])

_twist = IntegerSpinnerCommandInputHelper(
    'twistIpt',
    lm.sLng('Twist Count'),
    1,
    29,
    2,
    1)

_hole = IntegerSpinnerCommandInputHelper(
    'holeIpt',
    lm.sLng('Hole Count'),
    13,
    49,
    2,
    13)

_thick = ValueCommandInputHelper(
    'thickIpt',
    lm.sLng('Thickness'))

_info = TextBoxCommandInputHelper(
    'infoTxt',
    lm.sLng('Information'),
    '',
    10,
    True)

class MöbiusStripCore(apper.Fusion360CommandBase):

    def on_preview(self, command, inputs, args, input_values):
        global _point, _twist, _hole, _thick

        checkLst = [
            _twist.isRange(),
            _twist.isOdd(),
            _hole.isRange(),
            _hole.isOdd()
        ]

        if not all(checkLst):
            return

        MöbiusStripFactry.previewMöbiusStrip(
            _point.obj.selection(0).entity,
            _twist.obj.value,
            _hole.obj.value
        )

    def on_destroy(self, command, inputs, reason, input_values):
        pass

    def on_input_changed(self, command, inputs, changed_input, input_values):
        global _point, _twist, _hole, _thick, _info

        msg = []

        if not _twist.isRange():
            msg.append(f'{_twist.name} ' + lm.sLng('is out of scope!') + 
                f'({_twist.obj.minimumValue} - {_twist.obj.maximumValue})')

        if not _twist.isOdd():
            msg.append(f'{_twist.name} ' + lm.sLng('is an odd number only!'))


        if not _hole.isRange():
            msg.append(f'{_hole.name} ' + lm.sLng('is out of scope!') + 
                f'({_hole.obj.minimumValue} - {_hole.obj.maximumValue})')

        if not _hole.isOdd():
            msg.append(f'{_hole.name} ' + lm.sLng('is an odd number only!'))


        if not (_thick.obj.value > 0):
            msg.append(f'{_thick.name} ' + lm.sLng('should be a positive value!'))


        if len(msg) > 0:
            _info.obj.text = '\n'.join(msg)
        else:
            _info.obj.text = ''


        command.doExecutePreview()

    def on_execute(self, command, inputs, args, input_values):
        global _point, _twist, _hole, _thick
        thick = _thick.obj.value * 0.5
        MöbiusStripFactry.createMöbiusStrip(
            _point.obj.selection(0).entity,
            _twist.obj.value,
            _hole.obj.value,
            thick
        )

        _twist.initialValue =  _twist.obj.value
        _hole.initialValue = _hole.obj.value
        _thick.initialValue = adsk.core.ValueInput.createByString(_thick.obj.expression)

    def on_create(self, command, inputs):
        global _point, _twist, _hole, _thick, _info
        _point.register(inputs)
        _twist.register(inputs)
        _hole.register(inputs)
        _thick.register(inputs)
        _info.register(inputs)

    # expansions
    def on_validate(self, command, inputs, args, input_values):
        global _point, _twist, _hole, _thick, _info

        checkLst = [
            _twist.isOdd(),
            _hole.isOdd(),
            _thick.obj.value > 0
        ]

        if all(checkLst):
            args.areInputsValid = True
        else:
            args.areInputsValid = False

    # expansion
    def on_preselect(self, command, inputs, args, pre_input, input_values):
        pass